### ISSUE TITLE
How to call REST API with access only for Auth/Guest Users

### ISSUE THREAD

**Issue by aahmedfaraz** [aws-amplify/amplify-js#11975]
### Amplify CLI Version

12.4.0

### Background
- I created a web app on Next.js and added `Amplify Auth` in it. Every Auth function like `Auth.signup`, `Auth.signin` . . . etc, are running on client side, but as per my scenario I don't want to delete user on client side but inside a lambda, so that I can handle other related stuff as well inside the lambda. 
- For that, I want to create a REST API using Amplify CLI, which I will use to delete user and other related operations.

### What I did
- I created a REST API using Amplify CLI, and there is a lambda attached with it which will handle the operations, and while creating the API I added restrictions so that `only Auth/Guest user can access the api` and can only use `DELETE` method.
- Amplify created the API ✅, everything is fine till now.

### Problem
How to call this API using `API` object inside `aws-amplify`, I tried as per docs, it is giving `CORS Policy` error.

**Comment by aahmedfaraz** [comment]
Thanks, I also added the issue inside `amplify/cli`

**Comment by cwomack** [comment]
Hello, @aahmedfaraz 👋.  If you opened another issue inside the CLI repo, can you link it please so we know they are related?  

As for the REST API and CORS issues, can you share your `package.json` so we can see what dependencies and versions you have currently? Also was there any specific section of the docs or guide you were following for setting this up?

**Comment by ykethan** [comment]
https://github.com/aws-amplify/amplify-cli/issues/13189
duplicate issue on the CLI repository.

**Comment by ykethan** [comment]
@aahmedfaraz Does the Lambda function attached to the REST API path have a response with headers?

for example:
```
headers: {
       'Access-Control-Allow-Origin': '*',
       'Access-Control-Allow-Headers': '*',
     },
```

**Comment by aahmedfaraz** [comment]
No, I was not adding the wildcard with Headers, I did this now and it is working ✅ Thanks @ykethan

**Comment by aahmedfaraz** [comment]
One more thing, how to make sure the user which is running the lambda is Authorized? 

According to somewhere I studied it should have it in `event.requestContext.authorizer`, but there is no such variable available in the Lambda event, I just tried with providing a header on API call of `Authorization: 'Bearer <auth-jwt-token>`, and now when I hit the API, I got CORS again, these are the errors:

`Access to XMLHttpRequest at '<api-url>' from origin 'http://localhost:3000' has been blocked by CORS policy: Request header field authentication is not allowed by Access-Control-Allow-Headers in preflight response.`

and inside catch the `error.message` is `Network Error`

**Comment by josefaidt** [comment]
Hey @aahmedfaraz :wave:

> how to make sure the user which is running the lambda is Authorized?

by default when a REST API route is restricted using the CLI walkthrough, it is restricted and authenticated via IAM. The requests will be authorized before invoking the Lambda at the API Gateway level.

If you need to access the caller's identity inside the Lambda route handler, you will need to use [overrides to configure your route to authorize via Cognito User Pools](https://docs.amplify.aws/cli/restapi/override/#authorize-api-requests-with-cognito-user-pools), which you can then pass the bearer token and access the identity inside the handler.

>  I just tried with providing a header on API call of Authorization: 'Bearer <auth-jwt-token>, and now when I hit the API, I got CORS again, these are the errors:

Can you confirm in the API Gateway Console that there are CORS headers applied to default 4xx and 5xx responses? There was an issue in the past where subsequent updates to the REST API would remove these headers, and therefore when the handler/authorizer encounters an error you would see CORS errors in your client application rather than the actual error. For what it's worth Firefox will show the underlying error (if any), where Chrome swallows the response body and promotes the CORS error instead
